src/v/datalake/coordinator/coordinator.cc

```diff
@@ -162,23 +163,34 @@ coordinator::run_until_term_change(model::term_id term) {
             // TODO: apply table retention periodically too.
 
             auto updates = std::move(commit_res.value());
-            if (updates.empty()) {
-                // Nothing to replicate.
-                continue;
-            }
-            storage::record_batch_builder builder(
-              model::record_batch_type::datalake_coordinator, model::offset{0});
-            for (auto& update : updates) {
-                builder.add_raw_kv(
-                  serde::to_iobuf(mark_files_committed_update::key),
-                  serde::to_iobuf(std::move(update)));
+            if (!updates.empty()) {
+                storage::record_batch_builder builder(
+                  model::record_batch_type::datalake_coordinator,
+                  model::offset{0});
+                for (auto& update : updates) {
+                    builder.add_raw_kv(
+                      serde::to_iobuf(mark_files_committed_update::key),
+                      serde::to_iobuf(std::move(update)));
+                }
+                auto repl_res = co_await stm_->replicate_and_wait(
+                  term, std::move(builder).build(), as_);
+                if (repl_res.has_error()) {
+                    auto e = convert_stm_errc(repl_res.error());
+                    vlog(datalake_log.warn, "Replication failed {}", e);
+                    co_return e;
+                }
             }
-            auto repl_res = co_await stm_->replicate_and_wait(
-              term, std::move(builder).build(), as_);
-            if (repl_res.has_error()) {
-                auto e = convert_stm_errc(repl_res.error());
-                vlog(datalake_log.warn, "Replication failed {}", e);
-                co_return e;
+
+            // check if the topic has been deleted and we need to clean up
+            // topic state.
+            while (true) {
+                auto update_res = co_await update_lifecycle_state(t, term);
+                if (update_res.has_error()) {
+                    co_return update_res.error();
+                }
+                if (update_res.value() == ss::stop_iteration::yes) {
+                    break;
+                }
```

Comment:
Haven't fully thought through this, but I'm wondering if it makes sense to move this out of the coordinator, e.g. into a fiber of the controller leader to make RPC calls to the coordinator that only update the STM. Or at least, if we think about the lifecycle changes in that way, it's a forcing function to think about just the coordinator STM correctness.

Reply:
>Having it in a single fiber makes reasoning about it easier.

Yea I agree having a clear mental image of serialization makes it easier to grok. IMO that serialization is ideally just the STM apply fiber, if we can enumerate the different relevant states callers can expect of the STM.

My worry with having this be part of the loop is that the loop may run infrequently. I think in reconciling that with correctness in the STM, we've baked some implicit assumptions into the STM around revisions that seem kind of surprising on first pass.